planner/core/preprocess.go

```diff
@@ -1578,6 +1579,11 @@ func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
 	if p.LastSnapshotTS != 0 {
 		dom := domain.GetDomain(p.ctx)
 		p.InfoSchema, p.err = dom.GetSnapshotInfoSchema(p.LastSnapshotTS)
+		// If the statement contains local temporary table, we will return error.
+		if p.ctx.GetSessionVars().LocalTemporaryTableExists() {
+			p.err = errors.New("can not stale read temporary table")
```

Comment:
No, we should forbit stale read **on** the local temporary table...
Not forbiding the stale read feature when local temporary table **exists**.

Reply:
Should we allow snapshot normal table join local temporary table?

Reply:
> Should we allow snapshot normal table join local temporary table?

No... stale read on the local temporary table is not allowed
But on normal table is allowed.

**The snapshot ts works on the whole statement level, not on each table.**
You can use ts1 or table A join ts2 of table B ...

Reply:
Got it! 
I thought we can get ` snapshot normal table` after GetSnapshotInfoSchema when we query sql like: "snapshot normal table join local temporary table". But indeed even if LocalTemporaryTableExists, we might query only ` snapshot normal table` too.

Reply:
The codes in #25206 seems can cover both global and local temporary tables. So just add some tests for local temporary table should work. Also you can add some test cases to check stale read for normal table is also allowed when temporary tables exist.

Reply:
You can refer to `validCanReadTemporaryTable`.

Reply:
> The codes in #25206 seems can cover both global and local temporary tables. So just adding some cases for local temporary table should work. Also you can add some test cases to check stale read for normal table is also allowed when temporary tables exist.

adding test case is not enough, `GetSnapshotInfoSchema` can not get schema of local temporary table, which makes planner cannot see the local temporary table, return `table not exits` error.

Reply:
OK, I understand. So the `p.InfoSchema` should be assigned to the right infoschema with temporary table info. And please add below tests:

-  Stale read should be allowed when local temporary table exists.
- Because create/drop a temporary table do not commit the current transaction explicitly, so we should forbid these ddl operations in the stale read transaction.